lib/task_server.mli

```diff
@@ -0,0 +1,106 @@
+(* Task_server: Helper for interfaces of a particular shape
+   *
+   * This module is a helper module for interfaces that implement a Task interface
+   * with which asynchronous calls are implemented. In particular, it handles
+   * subtasks, cancellation, reporting of backtraces and a way of injecting
+   * failures at particular points known as 'cancellation triggers'
+*)
+
+type stringpair = string * string
+
+module type INTERFACE =
+sig
+  val service_name : string
+  exception Does_not_exist of stringpair
+  exception Cancelled of string
+  module Task :
+  sig
+    type id = string
+    type async_result
+    val rpc_of_async_result : async_result -> Rpc.t
+    val async_result_of_rpc : Rpc.t -> async_result
+    type completion_t = {
+      duration : float;
+      result : async_result option;
+    }
+    type state =
+        Pending of float
```

Comment:
I don't think it's ever non-zero, but it's very tricky to change this without breaking the interface between xapi, xenopsd and the storage backends.